tools/castle-editor/code/framedesign.pas

```diff
@@ -1970,6 +2027,39 @@ procedure TDesignFrame.ControlsTreeKeyUp(Sender: TObject; var Key: Word;
       ControlsTree.Selected.EditText;
 end;
 
+procedure TDesignFrame.ControlsTreeMouseUp(Sender: TObject;
+  Button: TLclMouseButton; Shift: TShiftState; X, Y: Integer);
+var
+  Sel: TComponent;
+begin
+  if (Button = TLclMouseButton.mbRight) then
```

Comment:
Instead of detecting in `OnMouseUp` event the right-click, it seems cleaner to assign the popup menu to `PopupMenu` property of the `ControlsTree` .

Then some stuff would be handled automatically. Detecting right-click in our code would not be necessary, setting `MenuTreeView.PopupComponent` would not be necessary, calling `MenuTreeView.PopUp` would not be necessary.

Reply:
Took me some time to find our conversation on the topic on 28 July 2020:

>    Instead of detecting in OnMouseUp event the right-click, it seems cleaner to assign the popup menu to PopupMenu property of the ControlsTree .

It throws SIGSEGVs for some reason, some variables are still uninitialized when PopupMenu is called.

> Hm, this is something I would investigate then more. The code will be simpler if we don't manually detect mbRight.

But it doesn't work :) I didn't investigate this bug too deeply, but AFAIR it seems like the menu items.visible isn't available at that moment (because the popup isn't shown yet, so there is no menu instance) and causes a SIGSEGV.

> Which would be solved by my suggestion with actions :slight_smile: You would then control actions, and no need to change menu items.

Reply:
I admit I'd like to require this change. That is, do not detect `mbRight`, instead use LCL `PopupMenu` property to assign the context menu to the control.

Manually detecting `mbRight` in `OnMouseUp` feels to me like a hackish way of invoking a popup menu. And e.g. you probably don't handle the "menu key" ( https://en.wikipedia.org/wiki/Menu_key ) this way. Making it cleaner would also mean you can remove `TLclMouseButton`.

You can use the `OnPopup` event to set initial enabled / visible properties, http://docs.embarcadero.com/products/rad_studio/delphiAndcpp2009/HelpUpdate2/EN/html/delphivclwin32/Menus_TPopupMenu_OnPopup.html . The menu should be fully initialized at this point (even more than in `TDesignFrame.ControlsTreeMouseUp` :) ).

Reply:
Ok, I have no idea what was wrong before, but now it just works. No SIGSEGVs whatsoever. All fixed, workarounds removed.